File: kotlin/kernel-test/src/main/kotlin/dev/modaal/duet/test/KernelTrace.kt
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

package dev.modaal.duet.test

import dev.modaal.duet.kernel.KernelClock
import dev.modaal.duet.kernel.LiveClock
import dev.modaal.duet.kernel.Reduced
import dev.modaal.duet.kernel.Store
import dev.modaal.duet.kernel.serialization.CanonicalJson
import dev.modaal.duet.kernel.serialization.CanonicalSerializers
import kotlin.time.Duration.Companion.nanoseconds
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.onCompletion
import kotlinx.coroutines.test.TestScope
import kotlinx.coroutines.yield
import kotlinx.serialization.KSerializer
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

// G3 — the kernel-trace apparatus, Kotlin side: the KMP flavor replays the SAME
// committed trace fixtures the Swift flavor records
// (swift/Tests/parity/fixtures/kernel-trace/), byte-identically, under
// coroutine virtual time. Contract: contracts/kernel-trace-v0.md. The Swift
// recorder is the corpus's writer of record; this side only verifies.

/** One observable kernel event, in trace order. Payload strings are canonical JSON. */
sealed interface KernelTraceEvent {
  data class Send(val action: String) : KernelTraceEvent
  data class Reduce(val action: String, val state: String) : KernelTraceEvent
  data class EffectStart(val payload: String, val id: String?) : KernelTraceEvent
  data class EffectAction(val action: String) : KernelTraceEvent
  data class EffectCancelled(val payload: String) : KernelTraceEvent
  data class EffectFinished(val payload: String) : KernelTraceEvent
  data class Advance(val nanoseconds: Long) : KernelTraceEvent
  data object Teardown : KernelTraceEvent
}

/** One canonical JSON line per event (jsonl — the fixture's byte form). */
val KernelTraceEvent.canonicalLine: String
  get() {
    val obj =
      when (this) {
        is KernelTraceEvent.Send -> buildJsonObject {
          put("event", "send")
          put("action", action)
        }
        is KernelTraceEvent.Reduce -> buildJsonObject {
          put("event", "reduce")
          put("action", action)
          put("state", state)
        }
        is KernelTraceEvent.EffectStart -> buildJsonObject {
          put("event", "effectStart")
          put("payload", payload)
          if (id == null) put("id", JsonNull) else put("id", id)
        }
        is KernelTraceEvent.EffectAction -> buildJsonObject {
          put("event", "effectAction")
          put("action", action)
        }
        is KernelTraceEvent.EffectCancelled -> buildJsonObject {
          put("event", "effectCancelled")
          put("payload", payload)
        }
        is KernelTraceEvent.EffectFinished -> buildJsonObject {
          put("event", "effectFinished")
          put("payload", payload)
        }
        is KernelTraceEvent.Advance -> buildJsonObject {
          put("event", "advance")
          put("nanoseconds", nanoseconds)
        }
        KernelTraceEvent.Teardown -> buildJsonObject { put("event", "teardown") }
      }
    return CanonicalJson.canonicalString(obj)
  }

/**
 * Records a kernel trace by wrapping a feature's reducer and handler around a
 * REAL [Store] on the test scope's virtual scheduler. Determinism is
 * structural on this platform (thinness notes vs the Swift twin):
 *
 * - **Lockstep is free**: a Flow's `emit` runs the store's collector — and
 *   therefore the send→reduce cycle — synchronously in the same coroutine, so
 *   delivery order in the trace IS reduce order with no wait loop.
 * - **The log needs no lock**: every append runs on the single-threaded test
 *   dispatcher (effect endings included — cancellation resumptions dispatch
 *   there too).
 * - Quiescence is the same signal as Swift: no new events across consecutive
 *   drain rounds (`runCurrent` + `yield` — the foreground-filter finding in
 *   DeterministicAsync.kt applies here too).
 *
 * Swift-flavor mirror: DuetTesting/KernelTrace.swift.
 */
@OptIn(ExperimentalCoroutinesApi::class)
class KernelTraceRecorder<S, A, P>(
  private val scope: TestScope,
  initialState: S,
  stateSerializer: KSerializer<S>,
  private val actionSerializer: KSerializer<A>,
  payloadSerializer: KSerializer<P>,
  reducer: (S, A) -> Reduced<S, P>,
  handler: (P, KernelClock) -> Flow<A>,
) {
  /** Virtual under the test scheduler — `advance` moves it. */
  val clock: KernelClock = LiveClock

  private val json = CanonicalSerializers.json
  private val events = mutableListOf<KernelTraceEvent>()

  private fun <T> canonical(serializer: KSerializer<T>, value: T): String =
    CanonicalJson.canonicalString(json.encodeToJsonElement(serializer, value))

  val store: Store<S, A, P> =
    Store(
      initialState = initialState,
      reducer = { state, action ->
        val reduced = reducer(state, action)
        events.add(
          KernelTraceEvent.Reduce(
            action = canonical(actionSerializer, action),
            state = canonical(stateSerializer, reduced.state)))
        reduced
      },
      handler = { payload ->
        val payloadCanonical = canonical(payloadSerializer, payload)
        // Synchronous with effect start (the Store invokes handlers inside
        // execute — the trace-pinned flavor-invariant ordering).
        events.add(KernelTraceEvent.EffectStart(payload = payloadCanonical, id = null))
        handler(payload, clock)
          .let { upstream ->
            flow {
              upstream.collect { action ->
                events.add(
                  KernelTraceEvent.EffectAction(action = canonical(actionSerializer, action)))
                emit(action) // the store's collector reduces synchronously in here
              }
            }
          }
          .onCompletion { cause ->
            if (cause is CancellationException) {
              events.add(KernelTraceEvent.EffectCancelled(payload = payloadCanonical))
            } else {
              events.add(KernelTraceEvent.EffectFinished(payload = payloadCanonical))
            }
          }
      },
      scope = scope.backgroundScope,
    )

  /** Scripted send: logs the stimulus, sends, then settles to quiescence. */
  suspend fun send(action: A) {
    events.add(KernelTraceEvent.Send(action = canonical(actionSerializer, action)))
    store.send(action)
    settleQuiescent()
  }

  suspend fun advance(byNanoseconds: Long) {
    events.add(KernelTraceEvent.Advance(nanoseconds = byNanoseconds))
    scope.testScheduler.advanceTimeBy(byNanoseconds.nanoseconds)
    settleQuiescent()
  }

  suspend fun teardown() {
    events.add(KernelTraceEvent.Teardown)
    store.teardown()
    settleQuiescent()
  }

  /**
   * Drains until the trace is stable for [stableRounds] consecutive rounds
   * (bounded): pending clocked sleeps produce no events until advanced, so
   * stability — not idleness — is the quiescence signal.
   */
  private suspend fun settleQuiescent(stableRounds: Int = 3, maxRounds: Int = 200) {
    var streak = 0
    var rounds = 0
    var lastCount = events.size
    while (streak < stableRounds && rounds < maxRounds) {
      scope.testScheduler.runCurrent()
      yield()
      rounds += 1
      val count = events.size
      if (count == lastCount) {
        streak += 1
      } else {
        streak = 0
        lastCount = count
      }
    }
  }

  /** The fixture's byte form: one canonical JSON line per event + trailing newline. */
  fun canonicalTrace(): String =
    events.joinToString(separator = "\n", postfix = "\n") { it.canonicalLine }
}

File: swift/Sources/DuetTesting/KernelTrace.swift
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

import Duet
import Foundation

// G3 (spec 18 §2.1) — kernel-trace fixtures: the corpus methodology extended to
// the KERNEL itself. The contract-observable runtime rules (synchronous
// send→reduce, effect start order, cancel-by-id, teardown-cancels-everything,
// reentrancy queueing, virtual-time ordering) are recorded as replayable traces
// both kernel flavors must reproduce under virtual time — machine-verifying
// what the twin TestStores pin per platform by assertion. Contract:
// contracts/kernel-trace-v0.md (v0 — frozen at F3 exit with the Kotlin traces).

/// One observable kernel event, in trace order. Payload strings are canonical
/// JSON (contracts/serialization.md) so the trace itself is byte-comparable.
public enum KernelTraceEvent: Equatable {
  /// The test scripted a send (the stimulus boundary — everything below is the
  /// kernel's response).
  case send(action: String)
  /// The reducer ran for `action`; `state` is the canonical post-reduce state.
  /// Ordering receipt: every `send`/`effectAction` is followed by its `reduce`
  /// BEFORE any nested reduce — the synchronous send→reduce + queued
  /// reentrancy rules, visible as a flat (never nested) event stream.
  case reduce(action: String, state: String)
  /// The handler was invoked for `payload` (the effect's task started).
  case effectStart(payload: String, id: String?)
  /// An effect's stream yielded `action` back into the store.
  case effectAction(action: String)
  /// An effect's stream terminated by cancellation (cancel-by-id, a
  /// same-id restart, or teardown).
  case effectCancelled(payload: String)
  /// An effect's stream ran to completion.
  case effectFinished(payload: String)
  /// The test advanced the virtual clock.
  case advance(nanoseconds: UInt64)
  /// The test tore the store down.
  case teardown
}

extension KernelTraceEvent {
  /// One canonical JSON line per event (jsonl — the fixture's byte form).
  public var canonicalLine: String {
    func line(_ object: [String: Any]) -> String {
      (try? CanonicalJSON.canonicalString(fromJSONObject: object)) ?? "{\"error\":true}"
    }
    switch self {
    case let .send(action):
      return line(["event": "send", "action": action])
    case let .reduce(action, state):
      return line(["event": "reduce", "action": action, "state": state])
    case let .effectStart(payload, id):
      return line(["event": "effectStart", "payload": payload, "id": id ?? NSNull()])
    case let .effectAction(action):
      return line(["event": "effectAction", "action": action])
    case let .effectCancelled(payload):
      return line(["event": "effectCancelled", "payload": payload])
    case let .effectFinished(payload):
      return line(["event": "effectFinished", "payload": payload])
    case let .advance(nanoseconds):
      return line(["event": "advance", "nanoseconds": nanoseconds])
    case .teardown:
      return line(["event": "teardown"])
    }
  }
}

/// Records a kernel trace by wrapping a feature's reducer and handler around a
/// REAL `Store` under a `TestClock`. Effect-side events are appended through a
/// lock (stream termination fires from the cancelling context); scripted steps
/// settle before returning so traces are deterministic by construction.
@MainActor
public final class KernelTraceRecorder<
  State: Equatable & Encodable, Action: Equatable & Encodable, Payload: Equatable & Encodable
> {
  public let clock = TestClock()
  public private(set) var store: Store<State, Action, Payload>!

  private let log = EventLog()

  public init(
    initialState: State,
    reducer: @escaping (inout State, Action) -> [Effect<Payload>],
    handler: @escaping (Payload, TestClock) -> AsyncStream<Action>
  ) {
    let log = self.log
    let clock = self.clock
    self.store = Store(
      initialState: initialState,
      reducer: { state, action in
        let effects = reducer(&state, action)
        log.append(.reduce(action: canonical(action), state: canonical(state)))
        return effects
      },
      handler: { payload in
        log.append(.effectStart(payload: canonical(payload), id: nil))
        let upstream = handler(payload, clock)
        // Re-wrap so yields and endings are observable in trace order. Two
        // determinism rules make the trace canonical rather than scheduling-
        // shaped: (1) LOCKSTEP forwarding — after yielding an action, wait for
        // its reduce to land before forwarding the next, so delivery order in
        // the trace IS reduce order (the contract's meaningful ordering; a
        // free-running forwarder races the store's consuming task); (2) the
        // ending is logged in ONE place, the wrapper task's tail, after the
        // last lockstep round.
        return AsyncStream { continuation in
          let task = Task { @MainActor in
            for await action in upstream {
              // Canonicalize BEFORE the yield sends the value away (region
              // isolation): the lockstep loop only touches the Sendable string.
              let canonicalAction = canonical(action)
              let mark = log.count
              log.append(.effectAction(action: canonicalAction))
              // The element is forwarded exactly once and never touched again
              // — the lockstep loop below reads only the canonical string.
              // Swift 6.4 proves that unaided; the GA line's 6.3 sending
              // analysis (Xcode 26.x) merges the element's region with the
              // task's captures and reports a phantom use-after-send at the
              // yield. The unsafe binding suppresses exactly that false
              // positive so DuetTesting keeps compiling on GA toolchains —
              // the adopter floor. Remove when the floor moves past 6.3.
              nonisolated(unsafe) let transferred = action
              continuation.yield(transferred)
              var budget = 10_000
              while !Task.isCancelled && budget > 0
                && !log.containsReduce(of: canonicalAction, atOrAfter: mark) {
                await Task.yield()
                budget -= 1
              }
            }
            if Task.isCancelled {
              log.append(.effectCancelled(payload: canonical(payload)))
            } else {
              log.append(.effectFinished(payload: canonical(payload)))
            }
            continuation.finish()
          }
          continuation.onTermination = { _ in
            task.cancel()
          }
        }
      })
  }

  /// Scripted send: logs the stimulus, sends, then settles to QUIESCENCE so
  /// every kernel response that can land does before the next step — a fixed
  /// yield count is not deterministic under host load; "no new events across
  /// consecutive drain rounds" is.
  public func send(_ action: Action) async {
    log.append(.send(action: canonical(action)))
    store.send(action)
    await settleQuiescent()
  }

  public func advance(byNanoseconds delta: UInt64) async {
    log.append(.advance(nanoseconds: delta))
    await clock.advance(byNanoseconds: delta)
    await settleQuiescent()
  }

  public func teardown() async {
    log.append(.teardown)
    store.teardown()
    await settleQuiescent()
  }

  /// Drains until the trace is stable for `stableRounds` consecutive
  /// mega-yield rounds (bounded): pending clocked sleeps produce no events
  /// until advanced, so stability — not idleness — is the quiescence signal.
  private func settleQuiescent(stableRounds: Int = 3, maxRounds: Int = 200) async {
    var streak = 0
    var rounds = 0
    var lastCount = log.count
    while streak < stableRounds && rounds < maxRounds {
      await Task.megaYield()
      rounds += 1
      let count = log.count
      if count == lastCount {
        streak += 1
      } else {
        streak = 0
        lastCount = count
      }
    }
  }

  public var events: [KernelTraceEvent] { log.snapshot() }

  /// The fixture's byte form: one canonical JSON line per event + trailing \n.
  public func canonicalTrace() -> String {
    log.snapshot().map(\.canonicalLine).joined(separator: "\n") + "\n"
  }
}

private func canonical<T: Encodable>(_ value: T) -> String {
  (try? CanonicalJSON.canonicalString(of: value)) ?? "<encoding-error>"
}

/// Lock-guarded event log — effect terminations append from the cancelling
/// context; everything else from the main actor.
private final class EventLog: @unchecked Sendable {
  private let lock = NSLock()
  private var events: [KernelTraceEvent] = []

  func append(_ event: KernelTraceEvent) {
    lock.lock()
    events.append(event)
    lock.unlock()
  }

  func snapshot() -> [KernelTraceEvent] {
    lock.lock()
    defer { lock.unlock() }
    return events
  }

  var count: Int {
    lock.lock()
    defer { lock.unlock() }
    return events.count
  }

  /// Lockstep support: has `action`'s reduce landed at or after `index`?
  func containsReduce(of action: String, atOrAfter index: Int) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    guard index < events.count else { return false }
    return events[index...].contains { event in
      if case let .reduce(reduced, _) = event { return reduced == action }
      return false
    }
  }
}
